In [24]:
import numpy as np
import pandas as pd
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Activation, Dense
from sklearn.utils import shuffle
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder


In [25]:
data = pd.read_csv("./Datasets/Iris.csv",names=['sepal length','sepal width','petal length','petal width','target'],skipinitialspace=True, skiprows=1, engine="python")
features = ['sepal length','sepal width','petal length','petal width']

main_input = data.loc[:,features].values
output = data.loc[:,['target']].values

main_input = np.array(main_input)
output = np.array(output)


encoder = LabelEncoder()
output = encoder.fit_transform(output)

output = pd.get_dummies(output).values



/home/jp/anaconda3/lib/python3.7/site-packages/sklearn/preprocessing/_label.py:251: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [147]:
def get_model():

    model = Sequential()
    model.add(Dense(10, input_shape=(4,),activation="relu"))
    model.add(Dense(8,activation="relu"))
    model.add(Dense(6,activation="relu"))
    model.add(Dense(3,activation="softmax"))
    
    return model;


In [148]:
X_train, X_test, y_train, y_test = train_test_split(main_input, output, test_size=0.2,shuffle=True)
print(np.shape(X_train))
print(np.shape(y_train))

(120, 4)
(120, 3)


In [149]:
model = get_model()
adam = tf.keras.optimizers.SGD(learning_rate=0.005)
model.compile(adam,'categorical_crossentropy',metrics=['accuracy'])

In [150]:
model.fit(x=X_train,y=y_train,epochs=100,shuffle=True,verbose=0) #do verbose 1 or 2 to see training

In [151]:
model.evaluate(X_test,y_test)

30/30 [==============================] - 0s 3ms/sample - loss: 0.9787 - accuracy: 0.5333


[0.9786794185638428, 0.53333336]

With Adam(lr=0.005)

with relu  loss: 0.1475 - accuracy: 0.9667<br>
with elu   loss: 0.1028 - accuracy: 0.9333<br>
with tanh  loss: 0.0915 - accuracy: 0.9667<br>

With Adam(lr=0.05)

with relu  loss: 0.0029 - accuracy: 1.0000<br>
with elu   loss: 0.0409 - accuracy: 1.0000<br>
with tanh  loss: 0.1215 - accuracy: 0.9333<br>


With SGD(lr=0.05)

with relu  loss: 0.0408 - accuracy: 1.0000<br>
with elu   loss: 0.0770 - accuracy: 0.9667<br>
with tanh  loss: 0.2518 - accuracy: 0.9000<br>

With SGD(lr=0.005)

with relu  loss: 0.9787 - accuracy: 0.5333<br>
with elu   loss: 0.5909 - accuracy: 0.6667<br>
with tanh  loss: 0.7112 - accuracy: 0.4667<br>
